### DISCUSSION TITLE
Type generation: create Block types individually

### DISCUSSION THREAD

**Discussion by AlessioGr** [payloadcms/payload#2502] [Feature Requests & Ideas]
**Update:** This feature request is no longer relevant, as it is now possible via the new `interfaceName` property: https://payloadcms.com/docs/typescript/generating-types#custom-field-interfaces

---

Currently, blocks in the generated types are all defined inline in the field which is of type `blocks`. E.g.:
![Code 2023-04-16 at 16 14 58@2x](https://user-images.githubusercontent.com/70709113/232318994-35d22d51-6b7f-4248-acf5-02323f47922d.jpg)

My proposal is to save the blocks into their own types:
![Code 2023-04-16 at 16 15 21@2x](https://user-images.githubusercontent.com/70709113/232319017-f1a2ae14-6f4e-479a-8778-e84e10df8d13.jpg)

Then save the blocks union into its own type:
![Code 2023-04-16 at 16 15 45@2x](https://user-images.githubusercontent.com/70709113/232319055-0fc76add-ca9a-40a7-8351-2edf93cc131e.jpg)

And use _**that**_ in the collection:
![Code 2023-04-16 at 16 16 17@2x](https://user-images.githubusercontent.com/70709113/232319094-71408933-3929-4c35-a2b7-2aee0a801fd2.jpg)

## Why?

This allows me to re-use the individual Block types in my frontend, which is very useful when looping through each block and trying to transform it into html for my front-end.

It would also reduce the size of the generated types if 2 components are using the same blocks ;)

**Comment by denolfe** [comment]
We've definitely talked about this internally. Seems like it would be a great thing to update.

**Comment by cloetensbrecht** [comment]
Also waiting for this separate block types

**Comment by cloetensbrecht** [comment]
A current workaround can be:

```typescript
import type {Page} from './payload-types'

type Blocks = Page['blocks'][number]
// export each available block
// ex:
export type VideoBlock = Extract<Blocks, {blockType: 'video'}>
export type ImageBlock = Extract<Blocks, {blockType: 'image'}>
```

**Comment by AlessioGr** [comment]
I'm closing this discussion, as what I requested is now possible with the `interfaceName` property: https://payloadcms.com/docs/typescript/generating-types#custom-field-interfaces

Giving a Blocks field an `interfaceName` now generates a re-usable type for that block
